File: go/deliberation/pkg/deliberation/adjudicator_test.go
package deliberation

import (
	"context"
	"errors"
	"testing"
)

func mkVoice(name string, v Verdict, c float64) Voice {
	var ab *string
	if v == VerdictAbstain {
		s := "x"
		ab = &s
	}
	return Voice{Specialist: name, Verdict: v, Confidence: c, Rationale: "", Citations: []Citation{}, AbstainedBecause: ab}
}

func TestAdjudicateAllAbstain(t *testing.T) {
	r := Adjudicate(context.Background(), []Voice{
		mkVoice("a", VerdictAbstain, 0),
		mkVoice("b", VerdictAbstain, 0),
	}, AdjudicateOptions{})
	if r.Verdict != VerdictHold {
		t.Fatalf("expected hold, got %s", r.Verdict)
	}
	if r.Confidence != 0 {
		t.Errorf("confidence expected 0, got %f", r.Confidence)
	}
}

func TestAdjudicateUnanimousBuy(t *testing.T) {
	r := Adjudicate(context.Background(), []Voice{
		mkVoice("regime", VerdictBuy, 0.8),
		mkVoice("aletheia", VerdictBuy, 0.7),
	}, AdjudicateOptions{})
	if r.Verdict != VerdictBuy {
		t.Fatalf("expected buy, got %s", r.Verdict)
	}
	if r.Confidence != 1 {
		t.Errorf("confidence expected 1, got %f", r.Confidence)
	}
	if r.DissentScore != 0 {
		t.Errorf("dissent expected 0, got %f", r.DissentScore)
	}
}

func TestAdjudicateDissentEscalates(t *testing.T) {
	brainCalled := false
	r := Adjudicate(context.Background(), []Voice{
		mkVoice("regime", VerdictBuy, 0.7),
		mkVoice("aletheia", VerdictSell, 0.6),
	}, AdjudicateOptions{
		Brain: func(ctx context.Context, prompt string) (string, error) {
			brainCalled = true
			return `{"verdict":"hold","confidence":0.5,"explanation":"split"}`, nil
		},
	})
	if !brainCalled {
		t.Fatal("expected brain to be called on dissent")
	}
	if !r.BrainEscalated {
		t.Error("expected BrainEscalated true")
	}
	if r.DissentScore < DefaultDissentThreshold {
		t.Errorf("dissent_score %f below threshold %f", r.DissentScore, DefaultDissentThreshold)
	}
}

func TestAdjudicateBrainFailureGraceful(t *testing.T) {
	r := Adjudicate(context.Background(), []Voice{
		mkVoice("a", VerdictBuy, 0.7),
		mkVoice("b", VerdictSell, 0.7),
	}, AdjudicateOptions{
		Brain: func(ctx context.Context, prompt string) (string, error) {
			return "", errors.New("cascade down")
		},
	})
	if !r.BrainEscalated {
		t.Error("expected BrainEscalated true even on failure")
	}
	if r.BrainSynthesis != nil {
		t.Error("expected nil BrainSynthesis on failure")
	}
	if r.Verdict != VerdictBuy {
		t.Errorf("expected weighted-majority buy, got %s", r.Verdict)
	}
}

File: node/deliberation/test/adjudicator.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { adjudicate, DEFAULT_DISSENT_THRESHOLD } from '../src/adjudicator.mjs';

const v = (specialist, verdict, confidence) => ({
  specialist, verdict, confidence, rationale: '', citations: [],
  abstained_because: verdict === 'abstain' ? 'x' : null,
});

test('adjudicate: all abstain → hold/0', async () => {
  const r = await adjudicate([
    v('regime', 'abstain', 0),
    v('aletheia', 'abstain', 0),
  ]);
  assert.equal(r.verdict, 'hold');
  assert.equal(r.confidence, 0);
  assert.equal(r.brain_escalated, false);
});

test('adjudicate: unanimous buy → buy with full conf share', async () => {
  const r = await adjudicate([
    v('regime', 'buy', 0.8),
    v('aletheia', 'buy', 0.7),
  ]);
  assert.equal(r.verdict, 'buy');
  assert.equal(r.confidence, 1);
  assert.equal(r.dissent_score, 0);
});

test('adjudicate: split buy/sell triggers dissent escalation', async () => {
  let brainCalled = false;
  const r = await adjudicate(
    [
      v('regime', 'buy', 0.7),
      v('aletheia', 'sell', 0.6),
    ],
    {
      brain: async (prompt) => {
        brainCalled = true;
        assert.match(prompt, /DISSENT/);
        return JSON.stringify({ verdict: 'hold', confidence: 0.5, explanation: 'split' });
      },
    },
  );
  assert.equal(brainCalled, true);
  assert.equal(r.brain_escalated, true);
  assert.ok(r.dissent_score >= DEFAULT_DISSENT_THRESHOLD);
});

test('adjudicate: minor dissent below threshold → no brain', async () => {
  const r = await adjudicate(
    [
      v('a', 'buy', 0.9),
      v('b', 'buy', 0.8),
      v('c', 'sell', 0.1),
    ],
    { brain: async () => 'should-not-be-called' },
  );
  assert.equal(r.brain_escalated, false);
  assert.equal(r.verdict, 'buy');
});

test('adjudicate: brain failure during escalation does not throw', async () => {
  const r = await adjudicate(
    [
      v('a', 'buy', 0.7),
      v('b', 'sell', 0.7),
    ],
    {
      brain: async () => { throw new Error('cascade down'); },
    },
  );
  assert.equal(r.brain_escalated, true);
  assert.equal(r.brain_synthesis, null);
  assert.equal(r.verdict, 'buy');  // weighted-majority still wins
});

test('adjudicate: dissent log identifies opposers', async () => {
  const r = await adjudicate([
    v('regime', 'buy', 0.7),
    v('aletheia', 'sell', 0.6),
    v('empirical', 'hold', 0.5),
  ]);
  assert.ok(r.dissent_log.some(d => d.includes('aletheia')));
  assert.ok(!r.dissent_log.some(d => d.includes('regime')));
});

test('adjudicate: hold-only voices → hold leader', async () => {
  const r = await adjudicate([
    v('a', 'hold', 0.6),
    v('b', 'hold', 0.7),
  ]);
  assert.equal(r.verdict, 'hold');
});

test('adjudicate: custom dissent threshold respected', async () => {
  const r = await adjudicate(
    [
      v('a', 'buy', 0.6),
      v('b', 'sell', 0.4),
    ],
    { dissentThreshold: 0.10, brain: async () => null },
  );
  assert.equal(r.brain_escalated, true);
});
